### PR TITLE
Unify engine error to reject blocks

### PR DESCRIPTION
**Pull request by sorpaas**
Currently in our code, we have several places that deal with engine errors with different behavior. On engine error:

* If `on_new_block` fails, blocks are *rejected* if it's imported externally.
* If `on_new_block` fails, the client *panics* if it's created internally.
* If `on_close_block` fails, the error is *ignored* and we print a warning.

I think the *rejected* behavior is the most reasonable. So this PR changes all three cases to rejecting blocks.

* If we ignore the error and only print a warning, it might be ignored by the node operator as well. It may end up making the chain not having the same best block, which is bad.
* In some cases, a consensus failure might be recoverable. For example, we might have a certain combination of transactions that cause `on_close_block` to fail on a consensus smart contract. In that case, if *rejected* behavior is used, node operators may be able to push/replace transactions and cause the current node to create a new valid block, without bringing down the node.

### CODE REVIEW

File: ethcore/src/miner/miner.rs [hunk 1]
```diff
@@ -349,11 +349,15 @@ impl Miner {
...
-					chain.prepare_open_block(
+
+					match chain.prepare_open_block(
 						params.author,
 						params.gas_range_target,
 						params.extra_data,
-					)
+					) {
+						Ok(block) => block,
+						Err(_) => return None,
```
Comment (tomusdrw) [inline ethcore/src/miner/miner.rs]: Should we print a warning here?

File: ethcore/src/miner/miner.rs [hunk 2]
```diff
@@ -480,15 +484,21 @@ impl Miner {
 		let elapsed = block_start.elapsed();
 		debug!(target: "miner", "Pushed {} transactions in {} ms", tx_count, took_ms(&elapsed));
 
-		let block = open_block.close();
+		let block = match open_block.close() {
+			Ok(block) => block,
+			Err(_) => {
+				warn!(target: "miner", "Closing the block failed due to an engine error. Please check your chain specifications and consensus smart contracts.");
```
Comment (tomusdrw) [inline ethcore/src/miner/miner.rs]: Maybe display the original error as well? Will give some additional input for debugging.

File: ethcore/src/miner/miner.rs [hunk 3]
```diff
@@ -714,8 +724,14 @@ impl Miner {
...
 			// | Make sure to release the locks before calling that method.             |
 			// --------------------------------------------------------------------------
-			let (block, original_work_hash) = self.prepare_block(client);
-			self.prepare_work(block, original_work_hash);
+			match self.prepare_block(client) {
+				Some((block, original_work_hash)) => {
+					self.prepare_work(block, original_work_hash)
+				},
+				None => {
+					prepare_new = false;
```
Comment (tomusdrw) [inline ethcore/src/miner/miner.rs]: Isn't that a bit misleading? The function should return `true` if new pending block had to be prepared. And in that case it had to be prepared, but the preparation failed.
Just thinking if it won't break any other assumptions, maybe at least update the function comment.

Comment (sorpaas) [reply ethcore/src/miner/miner.rs]: I added `BlockPreparationStatus` which distinguishes three different status.

I think previously I set `prepare_new = false` here is indeed not quite right. We only use the result for `prepare_pending_block` in `prepare_and_update_sealing`. If we don't have to prepare a new block, the function then call `update_sealing`. So if the block had to be prepared but failed, then we shouldn't continue the if statement. (`update_sealing` will also just call `prepare_block` again and early return if it finds out that the conditions does not hold.)

**Comment by 5chdn** [comment]
Removing relnotes flag as this is not relevant for users. Or is it?

**Comment by debris** [comment]
it is not

**Pull request merged**